File: 20-actors/matsurigoto/methods/test_sign_capability.py
#!/usr/bin/env python3
"""Tests for the R1.C sign/authority layer (matsurigoto 政, ADR-2606052300 + 2605231525)."""
from __future__ import annotations

import pathlib
import sys

HERE = pathlib.Path(__file__).resolve().parent
sys.path.insert(0, str(HERE))
sys.path.insert(0, str(HERE / "modules"))

import sign_capability as S  # noqa: E402
import tax_assess as T       # noqa: E402

COUNCIL = "did:web:etzhayyim.com:council:safe"
STATE = "did:web:gov.example:tax-authority"
AT = "2026-06-06T00:00:00Z"


def _unsigned():
    return T.assess_from_return(1_000_000, 0, "FLAT20.income")["receipt"]


def test_module_holds_no_key():
    assert S.SIGNER_HELD_PRIVATE_KEY is False


def test_server_side_signing_always_raises():
    """The structural no-server-key guarantee."""
    try:
        S.sign_server_side(_unsigned())
    except RuntimeError:
        return
    raise AssertionError("sign_server_side must raise (no-server-key)")


def test_principal_a_council_signs():
    signed = S.attach_external_proof(_unsigned(), signer_did=COUNCIL,
                                     authority_mode=":sovereign-governance",
                                     signature="0xSAFE", signed_at=AT)
    assert signed["proof"]["signer_did"] == COUNCIL
    assert "unsigned" not in signed["status"]
    assert S.verify_proof(signed) is True


def test_principal_b_state_signs_with_own_key():
    signed = S.attach_external_proof(_unsigned(), signer_did=STATE,
                                     authority_mode=":supplied-to-state",
                                     signature="0xSTATE", signed_at=AT)
    assert S.verify_proof(signed) is True


def test_principal_a_rejects_non_council_signer():
    """Sovereign acts must be signed by a Council organ, not an arbitrary did."""
    try:
        S.attach_external_proof(_unsigned(), signer_did=STATE,
                                authority_mode=":sovereign-governance",
                                signature="0xX", signed_at=AT)
    except ValueError:
        return
    raise AssertionError("sovereign act by a non-council signer must raise")


def test_principal_b_rejects_etzhayyim_holding_state_key():
    """etzhayyim never holds the adopting state's key — an etzhayyim did can't sign a state act."""
    try:
        S.attach_external_proof(_unsigned(), signer_did="did:web:etzhayyim.com:worker",
                                authority_mode=":supplied-to-state",
                                signature="0xX", signed_at=AT)
    except ValueError:
        return
    raise AssertionError("etzhayyim signing a supplied-to-state act must raise")


def test_empty_signature_refused():
    try:
        S.attach_external_proof(_unsigned(), signer_did=COUNCIL,
                                authority_mode=":sovereign-governance",
                                signature="", signed_at=AT)
    except ValueError:
        return
    raise AssertionError("empty signature must raise (matsurigoto mints none)")


def test_double_sign_refused():
    signed = S.attach_external_proof(_unsigned(), signer_did=COUNCIL,
                                     authority_mode=":sovereign-governance",
                                     signature="0xSAFE", signed_at=AT)
    try:
        S.attach_external_proof(signed, signer_did=COUNCIL,
                                authority_mode=":sovereign-governance",
                                signature="0xAGAIN", signed_at=AT)
    except ValueError:
        return
    raise AssertionError("re-signing an already-signed artifact must raise")


def test_tampered_payload_fails_verify():
    signed = S.attach_external_proof(_unsigned(), signer_did=COUNCIL,
                                     authority_mode=":sovereign-governance",
                                     signature="0xSAFE", signed_at=AT)
    signed["assessed_amount"] = 999_999   # tamper a SUBSTANTIVE field after signing
    assert S.verify_proof(signed) is False


def test_unsigned_artifact_does_not_verify():
    assert S.verify_proof(_unsigned()) is False


def _run():
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
        print(f"  ok  {fn.__name__}")
    print(f"\n{len(fns)}/{len(fns)} passed")
    return 0


if __name__ == "__main__":
    sys.exit(_run())

File: 20-actors/matsurigoto/methods/test_sign_capability.clj
;; test_sign_capability.clj — matsurigoto no-server-key sign layer: external-proof attach + verify
;; + principal A/B legitimacy + tamper detection. Run via `bb test:matsurigoto`. ADR-2606142300.
(ns matsurigoto.methods.test-sign-capability
  (:require [clojure.test :refer [deftest is testing run-tests]]
            [matsurigoto.methods.sign-capability :as sc]
            [matsurigoto.methods.modules.credential-issue :as ci]))

(def ^:private doc
  (:document (ci/issue-passport {:doc-number "L898902C3" :issuing-state "UTO" :nationality "UTO"
                                 :surname "ERIKSSON" :given-names "ANNA" :dob-yymmdd "740812"
                                 :sex "F" :expiry-yymmdd "120415" :subject-did "did:x"})))

(deftest no-server-key
  (testing "matsurigoto holds no key and signs nothing (ADR-2605231525)"
    (is (false? sc/signer-held-private-key))
    (is (thrown? Exception (sc/sign-server-side doc)))))

(deftest principal-a-council-sign-and-verify
  (testing "principal A — a Council organ signs externally; structure verifies"
    (let [signed (sc/attach-external-proof doc {:signer-did "did:web:etzhayyim.com:council:safe"
                                                :authority-mode :sovereign-governance
                                                :signature "0xCOUNCIL_SAFE_SIG" :signed-at "2026-06-06T00:00:00Z"})]
      (is (= "issued-signed" (:status signed)))             ; unsigned → signed
      (is (= "did:web:etzhayyim.com:council:safe" (get-in signed [:proof :signer-did])))
      (is (false? (:server-held-authority signed)))          ; still no operator key
      (is (true? (sc/verify-proof signed))))))

(deftest principal-b-state-key
  (testing "principal B — the adopting state's OWN (non-etzhayyim) key signs"
    (let [signed (sc/attach-external-proof doc {:signer-did "did:web:mof.go.jp:pki"
                                                :authority-mode :supplied-to-state
                                                :signature "0xSTATE_SIG" :signed-at "2026-06-06"})]
      (is (true? (sc/verify-proof signed))))))

(deftest illegitimate-signers-rejected
  (testing "principal A needs a Council did; principal B must NOT be an etzhayyim did"
    (is (thrown? Exception (sc/attach-external-proof doc {:signer-did "did:web:random.example"
                                                          :authority-mode :sovereign-governance
                                                          :signature "x" :signed-at "t"})))     ; A not council
    (is (thrown? Exception (sc/attach-external-proof doc {:signer-did "did:web:etzhayyim.com:council:safe"
                                                          :authority-mode :supplied-to-state
                                                          :signature "x" :signed-at "t"})))     ; B is etzhayyim
    (is (thrown? Exception (sc/attach-external-proof doc {:signer-did "did:x" :authority-mode :nonsense
                                                          :signature "x" :signed-at "t"})))))   ; unknown mode

(deftest attach-guards
  (testing "unsigned-on-arrival (G1) + a real external signature required"
    (is (thrown? Exception (sc/attach-external-proof doc {:signer-did "did:web:etzhayyim.com:council:s"
                                                          :authority-mode :sovereign-governance
                                                          :signature "" :signed-at "t"})))      ; empty sig
    (let [signed (sc/attach-external-proof doc {:signer-did "did:web:etzhayyim.com:council:s"
                                                :authority-mode :sovereign-governance
                                                :signature "0xSIG" :signed-at "t"})]
      (is (thrown? Exception (sc/attach-external-proof signed {:signer-did "did:web:etzhayyim.com:council:s"
                                                               :authority-mode :sovereign-governance
                                                               :signature "0xSIG2" :signed-at "t"}))))))  ; already signed

(deftest tamper-detection
  (testing "mutating a substantive field after signing breaks the payload digest"
    (let [signed (sc/attach-external-proof doc {:signer-did "did:web:etzhayyim.com:council:s"
                                                :authority-mode :sovereign-governance
                                                :signature "0xSIG" :signed-at "t"})]
      (is (true?  (sc/verify-proof signed)))
      (is (false? (sc/verify-proof (assoc signed :credential-subject {:id "did:forged"}))))   ; field tampered
      (is (false? (sc/verify-proof (dissoc signed :proof))))                                   ; no proof
      (is (false? (sc/verify-proof (assoc-in signed [:proof :signature] "")))))))               ; empty sig

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'matsurigoto.methods.test-sign-capability)]
    (System/exit (if (pos? (+ fail error)) 1 0))))
